api/include/opentelemetry/common/stltypes.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <opentelemetry/version.h>
+
+#ifndef HAVE_STDLIB_CPP
+/* Private implementation of standard C++20 types */
+#include "opentelemetry/nostd/stl/span.h"
+#include "opentelemetry/nostd/stl/string_view.h"
+#include "opentelemetry/nostd/stl/variant.h"
+#include "opentelemetry/nostd/stl/utility.h"
+#else
+/* Standard library implementation */
+#include <cstddef>
+#include <variant>
+#include <memory>
+#include <string_view>
```

Comment:
It doesn't violate the dynamic loading requirement for those who need 3rd party vendor library interop: those customers can build SDK with nostd implementation. It doesn't violate the static linking requirement either.

Reply:
> It doesn't violate the dynamic loading requirement for those who need 3rd party vendor library interop

To make this work, it requires that the users knows how each library was compiled (either with `std` or `nostd`). That might be hard, given that a user could install an OT-enabled library via `apt-get`. There also might be situations where a user gets two third-party library binaries, one compiled with `nostd` and the other one with `std`. Then there's no way those libraries can be used together with OT enabled.

I'm afraid it might be too much to ask from user: knowing how OT for each of their libraries was compiled. They only might find out when OT crashes their application.

Reply:
> To make this work, it requires that the users knows how each library was compiled

This is the case for Visual C++ STL today, where you cannot mix and match static libraries built for Release vs Debug, and even for the libraries prebuilt with vs2017 and vs2019. Such as Google Benchmark compiled with vs2019 is not gonna link to project built with vs2017.

What you are referring to (user could install an OT-enabled library via package manager) is not realistic.

Have we decided on planning to maintain prebuilt packages? For header-only implementation - it's be typical for a header (even in a standard library) to depend differently on what target you're building for. Such as "debug iterators and throw" on debug, or totally avoid iterators debugging on release.

Reply:
Re:
> when OT crashes their application.

We can make nostd default. I will provide a write-up that explains the 'pros' and 'cons', plus complete set of tests to verify that both implementations are interchangable:

- one goes with **nostd** if they are planning to release a prebuilt shared library interoperable with other exporters.

- one goes with **STL** (still same 100% **nostd** API that is aliased to Standard Library), to avoid the unnecessary perf penalties at translation layer. There are compliance reasons why some customers prefer absolutely no dynamic loading, and no dependencies on custom implementation of STL containers.